content/blog/2020-12-18-brilc.md

```diff
@@ -0,0 +1,248 @@
++++
+title="Struct Extension for Compiled Bril Programs"
+[[extra.authors]]
+name = "Mark Moeller"
+[extra]
+latex = true
+bio = """
+  Mark Moeller is a first-year graduate student in Computer Science at Cornell
+  interested in compilers and formal foundations of programming languages.
+"""
++++
+
+### Motivation: Garbage Collection for Compilers
+
+Building garbage collectors for compilers (as opposed to interpreters) is
+particularly tricky, since almost everything about garbage collection algorithms
+is informed by dynamic information. Assuming we are not in a virtual machine
+environment (such as Java, which can relegate details of garbage collection to
+the JVM), we must "bake in" statically a host of things to be checked or
+performed dynamically. This involves, at least, the following decisions:
+* What type of collection? While the two main choices here are tracing and
+  reference counting, we saw in the [Unified Theory of Garbage
+  Collection][unified], complete answers lie on a spectrum. This choice is somewhat
+  orthogonal to the compiler/interpreter issue.
+* When to collect? Certainly this decision will ideally be informed by dynamic
+  information about the heap state to minimize unnecessary pauses or overhead.
+  Without hardware or VM support, we may be limited in the choices of "hooks"
+  into the code to even choose from. For example, to be a drop-in replacement for
+  manual memory managment in C, one (naively at least) has only malloc calls to
+  work with as possible collection times.
```

Comment:
FWIW, the places in the code where the runtime system can step in to do stuff like GC are typically called "VM safepoints."